### PR TITLE
[security] add basic encryption features

### PR DESCRIPTION
**Pull request by dcramer**

### CODE REVIEW

**Comment by dcramer** [comment]
This pulls in the core encryption functionality from https://github.com/getsentry/sentry/pull/3485/files

**Comment by dcramer** [comment]
fwiw this isnt done -- im going to add a test suite around it

File: src/sentry/db/models/fields/bounded.py [hunk 1]
```diff
@@ -110,3 +110,6 @@ class BoundedBigAutoField(BoundedAutoField):
 add_introspection_rules([], ["^sentry\.db\.models\.fields\.bounded\.BoundedBigIntegerField"])
 add_introspection_rules([], ["^sentry\.db\.models\.fields\.bounded\.BoundedPositiveIntegerField"])
 add_introspection_rules([], ["^sentry\.db\.models\.fields\.pickle\.UnicodePickledObjectField"])
+add_introspection_rules([], ["^sentry\.db\.models\.fields\.encrypted\.EncryptedPickledObjectField"])
```
Comment (dcramer) [inline src/sentry/db/models/fields/bounded.py]: gonna rm -rf all the db changes here yet

**Comment by dcramer** [comment]
should be all good now

File: src/sentry/utils/encryption.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    def encrypt(self, value):
+        if self.default_scheme is None:
+            return value
+        value = smart_bytes(value)
+        scheme = self.schemes[self.default_scheme]
+        return b'{}${}'.format(self.default_scheme, scheme.encrypt(value))
+
+    def decrypt(self, value):
+        try:
+            enc_method, enc_data = value.split('$', 1)
```
Comment (dcramer) [inline src/sentry/utils/encryption.py]: I guess one concern is this won't be compatible with any values that have $ in them

I'm not sure if we care, but we could change the raise below to pass silently

File: src/sentry/utils/encryption.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+from django.conf import settings
+from django.utils.encoding import smart_bytes
+
+
+class EncryptionManager(object):
+    def __init__(self, schemes=()):
+        for key, value in schemes:
+            if not isinstance(key, basestring):
+                raise ValueError('Encryption scheme type must be a string. Value was: {!r}'.format(value))
+            if not hasattr(value, 'encrypt') and not hasattr(value, 'decrypt'):
```
Comment (bretthoerner) [inline src/sentry/utils/encryption.py]: `or`?

Comment (bretthoerner) [inline src/sentry/utils/encryption.py]: `or`?

File: src/sentry/conf/server.py [hunk 4]
```diff
@@ -1048,6 +1048,15 @@ def create_partitioned_queues(name):
 # unless you have altered all schemas first
 SENTRY_USE_BIG_INTS = False
 
+# Encryption schemes available to Sentry. You should *never* remove form this
+# list until the key is no longer used in the database. The first listed
+# implementation is considered the default and will be used to encrypt all
+# values (as well as re-encrypt data when its re-saved).
+SENTRY_ENCRYPTION_SCHEMES = (
```
Comment (bretthoerner) [inline src/sentry/conf/server.py]: `s/form/from/` and `s/its/it's/`

File: src/sentry/utils/encryption.py (hunk 2 above)
Comment (bretthoerner) [inline src/sentry/utils/encryption.py]: When you say values you just mean the `enc_method` name right? That seems fine to me, maybe we could blow up in `__init__` if any scheme keys us a `$`?

**Review by bretthoerner — DISMISSED** [review]

**Comment by dcramer** [comment]
should be all set now

File: src/sentry/utils/encryption.py (hunk 2 above)
Comment (dcramer) [inline src/sentry/utils/encryption.py]: i mean in the value for unencrypted legacy values (e.g. the value is $100 or something like that)

**Review by bretthoerner — APPROVED** [review]

**Pull request merged**